### DISCUSSION TITLE
shopify app not working on localhost (Osiset\ShopifyApp\Exceptions\MissingShopDomainException)

### DISCUSSION THREAD

**Discussion by Ronak-maker** [osiset/laravel-shopify#686] [Q&A]
Hey, @osiset 
I have installed the fresh laravel 8 and laravel-shopify package for developing a Shopify app. I did the all setting according to the GitHub wiki. in Shopify, my app is working fine, but on localhost, I get the error "Osiset\ShopifyApp\Exceptions\MissingShopDomainException". can I run the app on localhost as well?

route.php file:- 
`Route::get('/','ProductsController@view')->middleware(['auth.shopify'])->name('home');`

products.blade.php file:-
```
@extends('shopify-app::layouts.default')

@section('content')
    <router-view></router-view>
@endsection

@section('scripts')
    @parent

    <script type="text/javascript">
        var AppBridge = window['app-bridge'];
        var actions = AppBridge.actions;
        var TitleBar = actions.TitleBar;
        var Button = actions.Button;
        var Redirect = actions.Redirect;
        var titleBarOptions = {
            title: 'Products',
        };
        var myTitleBar = TitleBar.create(app, titleBarOptions);
    </script>
@endsection
```
i have copy the defult.blade.php file in `views\vendor\shopify-app\layouts`.
I am on windows  and working on xampp. please help

**Comment by moonpetaldev** [comment]
This is because Laravel expects a shop parameter. Append `?shop=myshop.myshopify.com` to the URL which will take you through Shopify Auth to get a session. If you don't enforce App Bridge, you will be able to use the app locally without having to go through the Shopify Admin interface. Your session may expire and you'll get the error again, just repeat the process. Hope this helps!

**Comment by Ronak-maker** [comment]
I know I have to setup some kind of certificate for that. But dont know how to setup “Laradock” or “Docker” in the pc.

**Reply by moonpetaldev** [reply]
Try taking a look at valet-windows - https://github.com/cretueusebiu/valet-windows

Also, it would be better if you could reply in a comment in GitHub rather than via email as your email opens a new answer and can congest the thread. Hope this helps!

**Comment by maganius** [comment]
I have the same problem but in the ajax request, I can view the homepage but if i send an ajax request to backend laravel throw this exception. I use axios ajax.

I notice in the dev tools the ajax return a html trying redirect to shopify auth, but it not work because its an ajax.

its the console.log of data returned

```HTML
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <base target="_top">

        <title>Redirecting...</title>

        <script type="text/javascript">
            document.addEventListener('DOMContentLoaded', function () {
                var redirectUrl = URLHIDDEN;
                if (window.top == window.self) {
                    // If the current window is the 'parent', change the URL by setting location.href
                    window.top.location.href = redirectUrl;
                } else {
                    // If the current window is the 'child', change the parent's URL with postMessage
                    normalizedLink = document.createElement('a');
                    normalizedLink.href = redirectUrl;

                    data = JSON.stringify({
                        message: 'Shopify.API.remoteRedirect',
                        data: { location: redirectUrl },
                    });
                    window.parent.postMessage(data, "https://krakencfe.myshopify.com");
                }
            });
        </script>
    </head>
    <body>
    </body>
</html>
```

**Reply by moonpetaldev** [reply]
Have you authenticated your session with Shopify before sending the Ajax request?
You can check by hitting a route that has the `auth.shopify` middleware and adding a `?shop=mystore.myshopify.com` parameter at the end of the url

**Reply by maganius** [reply]
> Have you authenticated your session with Shopify before sending the Ajax request?
> You can check by hitting a route that has the `auth.shopify` middleware and adding a `?shop=mystore.myshopify.com` parameter at the end of the url

![image](https://user-images.githubusercontent.com/8250243/109176820-f3c35880-7765-11eb-9cd2-ae60c82dcfd3.png)

I think its authenticated because If I enter to homepage my shop name is displayed correctly, but when I start an ajax with axios it try redirect to oauth, check image.

![image](https://user-images.githubusercontent.com/8250243/109177193-4bfa5a80-7766-11eb-8b4f-80503969b0c8.png)

`tienda (shop)` is displayed correctly.

**Comment by gnikyt** [comment]
Shouldn't need to add the shop. If you're logged in and cookies are fine. I
think you'll need to send cookies with the ajax call possibly

**Reply by maganius** [reply]
> Shouldn't need to add the shop. If you're logged in and cookies are fine. I think you'll need to send cookies with the ajax call possibly
> […](#)
> On Thu., Feb. 25, 2021, 12:06 maganius, ***@***.***> wrote: Have you authenticated your session with Shopify before sending the Ajax request? You can check by hitting a route that has the auth.shopify middleware and adding a ?shop=mystore.myshopify.com parameter at the end of the url [image: image] <https://user-images.githubusercontent.com/8250243/109176820-f3c35880-7765-11eb-9cd2-ae60c82dcfd3.png> I think its authenticated because If I enter to homepage my shop name is displayed correctly, but when I start an ajax with axios it try redirect to oauth, check image. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#686 (reply in thread)](https://github.com/osiset/laravel-shopify/discussions/686#discussioncomment-406199)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AASO4OUGWGDYG3O7AFEWIHTTAZVBHANCNFSM4XK6Z65Q> .

its the cookies sent through axios:

`XSRF-TOKEN=eyJpdiI6IlVOUmp0ZzZEdGhsdFBmWlNwc3ptM2c9PSIsInZhbHVlIjoiZlc5eVpjUDkvUFhkWTV0NjgzMThDZFZ1T0ZKZWJjVzYxaFY2MWlJVE9lWUczcXVVdGgwejlQZitXa0s5RnFTNC9pU3dUbXJEakhLOFVORFhjWFIyR3YvR3FyenJjY1lEZTJHWG1SekwxQjZ2bEZFaGIzNXFaNXdlY0VhSXpadVYiLCJtYWMiOiJlMGU1Nzk4MWVjYzVlOGQ4NjQzMGNlNTVlZGQ0ZDk0ZDU3MzhhNDJiZTE1NzZkYjBjZGU3MGNlNDI5MjE2YWE3In0=; shopify_session=eyJpdiI6IlFjZ0szVmxLSFFCTERGV1pVZ1phemc9PSIsInZhbHVlIjoic2ZVVzZqOVN2Z2xpUWEyVHNBRUJlTkVwSmNnQU5FSS9KQ3h2OWhBRTAxV3NpQkRQWGo0NW5vNjU4Szc1MnE1cHVTbVNyVGtuNHpyNkR6U2U0VE14b0ZIOHprQmZieTRNS0ZqSXhkQmtDZ3JuYXE1Sk1qMmcvbGs0U01KNU5rY1QiLCJtYWMiOiJjOGVhM2RjYTVmNTEwZTYwMTIwYWM0MDQ2ZjYyNTg5MDdkZmVkNWE0NDEzNWQyNWJhZDJkZWYyZjgzYWIzNjM5In0=`

**Reply by maganius** [reply]
If I remove the shop param it return this error:

![image](https://user-images.githubusercontent.com/8250243/109178059-389bbf00-7767-11eb-8ffe-4f74143698ab.png)

its my api route:

```
Route::middleware(['auth.shopify'])->group(function () {
    Route::get('/orders', 'ApiController@orders');
    Route::post('/cfe/create/{orden_id}', 'ApiController@createCFE');
});
```

**Reply by maganius** [reply]
In the ajax request this `$this->shopSession->guest()` always `return true` i dont know why.

**Reply by maganius** [reply]
I figured it out... I moved all routes from routes/api.php to routes/web.php and it get working.

**Reply by gnikyt** [reply]
api.php doesn't have web middleware which contains the stuff for the cookies.

**Reply by maganius** [reply]
I added the web middleware multiple times and it not work. I try again put web middleware on api routes without luck. I dont know why it not work... anyway... I put the api routes on web routes and working for me its ok.

**Reply by gnikyt** [reply]
api routes have no web, so they have no cookie acces. You would need to implement a token to send, theres many articles on this for Laravel. But its fine in web as well.
